xds/src/main/java/io/grpc/xds/sds/SdsClient.java

```diff
@@ -0,0 +1,394 @@
+                .channelType(EpollDomainSocketChannel.class);
+      } else {
+        builder = NettyChannelBuilder.forTarget(targetUri);
+      }
+      builder = builder.usePlaintext();
+      if (channelExecutor != null) {
+        builder = builder.executor(channelExecutor);
+      }
+      ManagedChannel channel = builder.build();
+      return new SdsClient(sdsSecretConfig, node, watcherExecutor, channel, eventLoopGroup);
+    }
+
+    @VisibleForTesting
+    static String extractUdsTarget(ConfigSource configSource) {
+      checkNotNull(configSource, "configSource");
+      checkArgument(
+          configSource.hasApiConfigSource(), "only configSource with ApiConfigSource supported");
+      ApiConfigSource apiConfigSource = configSource.getApiConfigSource();
+      checkArgument(
+          ApiType.GRPC.equals(apiConfigSource.getApiType()),
+          "only GRPC ApiConfigSource type supported");
+      checkArgument(
+          apiConfigSource.getGrpcServicesCount() == 1,
+          "expecting exactly 1 GrpcService in ApiConfigSource");
+      GrpcService grpcService = apiConfigSource.getGrpcServices(0);
+      checkArgument(
+          grpcService.hasGoogleGrpc() && !grpcService.hasEnvoyGrpc(),
+          "only GoogleGrpc expected in GrpcService");
+      GoogleGrpc googleGrpc = grpcService.getGoogleGrpc();
+      // for now don't support any credentials
+      checkArgument(
+          !googleGrpc.hasChannelCredentials()
+              && googleGrpc.getCallCredentialsCount() == 0
+              && Strings.isNullOrEmpty(googleGrpc.getCredentialsFactoryName()),
+          "No credentials supported in GoogleGrpc");
+      String targetUri = googleGrpc.getTargetUri();
+      checkArgument(!Strings.isNullOrEmpty(targetUri), "targetUri in GoogleGrpc is empty!");
+      return targetUri;
+    }
+  }
+
+  private SdsClient(
+      SdsSecretConfig sdsSecretConfig,
+      Node node,
+      Executor watcherExecutor,
+      ManagedChannel channel,
+      EventLoopGroup eventLoopGroup) {
+    checkNotNull(sdsSecretConfig, "sdsSecretConfig");
+    checkNotNull(node, "node");
+    this.sdsSecretConfig = sdsSecretConfig;
+    this.clientNode = node;
+    this.watcherExecutor = watcherExecutor;
+    this.eventLoopGroup = eventLoopGroup;
+    checkNotNull(channel, "channel");
+    this.channel = channel;
+  }
+
+  /**
+   * Starts resource discovery with SDS protocol. This method should be the first one to be called
+   * and should be called only once.
+   */
+  synchronized void start() {
+    if (requestObserver == null) {
+      secretDiscoveryServiceStub = SecretDiscoveryServiceGrpc.newStub(channel);
+      responseObserver = new ResponseObserver();
+      requestObserver = secretDiscoveryServiceStub.streamSecrets(responseObserver);
+    }
+  }
+
+  /** Stops resource discovery. No method in this class should be called after this point. */
+  synchronized void shutdown() {
+    if (requestObserver != null) {
+      requestObserver.onCompleted();
```

Comment:
Not sure I agree the implementation looks okay without the guard. But this will be called only from one place so I will remove the "synchronized"

Reply:
the other two use of requestObserver is in synchronized block of itself except this. thus those synchronized ones doesn't really need to be synchronized since it is just submitting to the executor (most of executors are thread safe).

Reply:
actually this `synchronized` was to guard against multiple threads calling `shutdown`. But that is a highly unlikely (impossible) scenario so I am going to remove it.